b3sum/src/main.rs

```diff
@@ -177,8 +184,94 @@ fn read_key_from_stdin() -> Result<[u8; blake3::KEY_LEN]> {
     }
 }
 
+fn check_file(input: impl Read, mmap_disabled: bool) -> Result<u64> {
+    let hasher = blake3::Hasher::new();
+    let mut num_failed = 0;
+
+    let reader = std::io::BufReader::new(input);
+    for (line_num, line) in reader.lines().enumerate() {
+        // Read the line and trim it
+        let line: String = line.with_context(|| format!("line {}", line_num + 1))?;
+        let line = line.trim();
+
+        // Skip the line if it's a comment or empty
+        if line.starts_with('#') {
+            continue;
+        }
+        if line.len() == 0 {
+            continue;
+        }
+
+        // Split line by whitespace
+        let mut line = line.split_whitespace();
+
+        // Get the hash value and parse the hex value
+        let hash: &str = line.next().expect("Unexpected missing value");
+        let hash = hex::decode(hash)
+            .with_context(|| format!("error decoding hex on line {}", line_num + 1))?;
+
+        // Get the filename
+        let filename: &str = line
+            .next()
+            .ok_or_else(|| Error::msg(format!("missing filename on line {}", line_num + 1)))?;
+        let os_filename: std::ffi::OsString = filename.into();
+
+        // Hash the file
+        let hasher = hasher.clone();
+        let mut check_hash = Vec::<u8>::with_capacity(hash.len());
+        let check_reader = hash_file(&hasher, &os_filename, mmap_disabled)
+            .with_context(|| format!("filename: {}", filename))?;
+        check_reader
+            .take(hash.len() as u64)
+            .read_to_end(&mut check_hash)?;
+
+        // Check if hashes match.
+        // TODO: Should this be constant time?
+        if check_hash == hash {
```

Comment:
@oconnor663 , I'd like your opinion of if we should be doing a constant-time check for equality here.

Reply:
I don't see any reason not to, and it might matter in some very obscure scenario. Let's stick to constant-time. We could even do a constant-time comparison on the hex bytes directly, and save ourselves the trouble of parsing them.

Reply:
You would have to lowercase both strings in that case, which is about the same trouble as parsing the hex string into bytes and comparing there.

Reply:
Are we allowed to assume that the output we're parsing came from `b3sum`? If so, maybe we don't have to support uppercase hex if we don't emit any? But anyway, parsing hex to bytes is also fine with me.